modules/ismp/clients/tendermint/verifier/src/types.rs

```diff
@@ -0,0 +1,352 @@
+		}
+		// Validate verification options
+		self.verification_options.validate()?;
+		Ok(())
+	}
+
+	/// Create a frozen trusted state
+	pub fn freeze_at_height(mut self, height: u64) -> Self {
+		self.frozen_height = Some(height);
+		self
+	}
+
+	/// Check if the state is valid for a given height
+	pub fn is_valid_for_height(&self, height: u64) -> bool {
+		match self.frozen_height {
+			Some(frozen_height) => height < frozen_height,
+			None => true,
+		}
+	}
+
+	/// Update the finalized header hash
+	pub fn update_finalized_header_hash(&mut self, new_hash: [u8; 32]) {
+		self.finalized_header_hash = new_hash;
+	}
+
+	/// Get the finalized header hash
+	pub fn get_finalized_header_hash(&self) -> [u8; 32] {
+		self.finalized_header_hash
+	}
+
+	/// Get verification options
+	pub fn get_verification_options(&self) -> &VerificationOptions {
+		&self.verification_options
+	}
+
+	/// Update verification options
+	pub fn update_verification_options(&mut self, options: VerificationOptions) {
+		self.verification_options = options;
+	}
+}
+
+impl Default for TrustedState {
+	fn default() -> Self {
+		Self {
+			chain_id: "test-chain".to_string(),
+			height: 1,
+			timestamp: 0,
+			finalized_header_hash: [0u8; 32],
+			validators: Vec::new(),
+			next_validators: Vec::new(),
+			next_validators_hash: [0u8; 32],
+			trusting_period: 3600, // 1 hour default
+			verification_options: VerificationOptions::default(),
+			frozen_height: None,
+		}
+	}
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ConsensusProof {
+	/// Signed header containing the block header and commit
+	pub signed_header: SignedHeader,
+	/// Ancestry of signed headers from trusted block height to the latest signed header
+	pub ancestry: Vec<SignedHeader>,
+	/// Next validator set (optional)
+	pub next_validators: Option<Vec<Validator>>,
+}
+
+impl ConsensusProof {
+	pub fn new(
+		signed_header: SignedHeader,
+		ancestry: Vec<SignedHeader>,
+		next_validators: Option<Vec<Validator>>,
+	) -> Self {
+		Self { signed_header, ancestry, next_validators }
+	}
+
+	pub fn height(&self) -> u64 {
+		self.signed_header.header.height.value()
+	}
+
+	pub fn timestamp(&self) -> u64 {
+		self.signed_header.header.time.unix_timestamp() as u64
+	}
+
+	pub fn chain_id(&self) -> &str {
+		self.signed_header.header.chain_id.as_str()
+	}
+
+	/// Get the lowest height in the ancestry (trusted block height)
+	pub fn trusted_height(&self) -> u64 {
+		self.ancestry
+			.first()
+			.map(|header| header.header.height.value())
+			.unwrap_or_else(|| self.signed_header.header.height.value())
+	}
+
+	/// Get the highest height in the ancestry (before the latest signed header)
+	pub fn ancestry_highest_height(&self) -> u64 {
+		self.ancestry
+			.last()
+			.map(|header| header.header.height.value())
+			.unwrap_or_else(|| self.trusted_height())
+	}
+
+	/// Check if the ancestry is continuous (no gaps in heights)
+	pub fn is_ancestry_continuous(&self) -> bool {
```

Comment:
The ancestry check must start from the parent header hash present in the target header, and follow the chain back until the trusted, finalised hash is reached by hashing each signed header in the ancestry list.